pkg/tagset/hash_generator.go

```diff
@@ -135,3 +135,121 @@ func (g *HashGenerator) Hash(tb *HashingTagsAccumulator) uint64 {
 
 	return hash
 }
+
+// Hash2 combines the hashes of unique tags in l and r accumulators. Duplicate tags are removed, so
+// at the end tag each tag is present once in either l or r, but not both at the same time.
+//
+// First, duplicates are removed from l. Then duplicates are removed from r, including any tags that
+// are already present in l. Can move tags from one accumulator to another.
+func (g *HashGenerator) Hash2(l *HashingTagsAccumulator, r *HashingTagsAccumulator) (uint64, uint64) {
```

Comment:
My initial thoughts on this was to always keep orchestrator tags unchanged (as they don't change often) and update only metrics tags by removing tags that are both in metrics and orchestrator. Doing so allow orchestrator tags to be reused very often. It would be also possible to cache a `map[string]struct{}` containing the hashes of the orchestrator tags to make the implementation of this function efficient.
Why do you move tags between accumulators?

Reply:
Orchestrator tags may come from several different sources inside the tagger (two different origins, and separate "orchestrator" entity tags); and thus may, in theory, contain duplicates. Tags come with precomputed hashes from the tagger, so all we do is shift some integers around as we fill the internal hashset, the deduplication part is not expensive. We don't use a `map[string]struct{}` currently, as we need individual numeric hash values to construct the context key, and want to hash each tag once.

Moving only happens when the combined size of the two accumulators exceeds the size of our pre-allocated hashset. In this case the fallback behavior is to copy all tags to a single accumulator and `SortUniq`. This isn't expected to happen in practice as the size of the hashset is sufficiently high; we just didn't want to have any sort of artificial limitation on number of tags the agent can handle.